pkg/operatorsource/configuring_phase_reconciler.go

```diff
@@ -0,0 +1,120 @@
+package operatorsource
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/operator-framework/operator-marketplace/pkg/apis/marketplace/v1alpha1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	catalogsourceconfigPrefix = "opsrc"
+)
+
+// NewConfiguringPhaseReconciler returns a PhaseReconciler which reconciles
+// an OperatorSource object in "Configuring" phase.
+func NewConfiguringPhaseReconciler(datastore Datastore, kubeclient Kubeclient) PhaseReconciler {
+	return &configuringPhaseReconciler{
+		datastore:  datastore,
+		kubeclient: kubeclient,
+	}
+}
+
+// Given a name of OperatorSource object, this function returns the name
+// of the corresponding CatalogSourceConfig type object
+func getCatalogSourceConfigName(operatorsourceName string) string {
+	return fmt.Sprintf("%s-%s", catalogsourceconfigPrefix, operatorsourceName)
+}
+
+// configuringPhaseReconciler is used to reconcile an OperatorSource object which is in "Configuring" phase.
+//
+// Reconcile reconciles an object of OperatorSource type which is currently in "Configuring" phase.
+// It ensures that a corresponding CatalogSourceConfig object exists.
+// Upon success, it returns "Succeeded" as the next desired phase.
+// On error, the function returns "Failed" as the next desied phase, and Message reflects the error message.
+// If the corresponding CatalogSourceConfig object already exists, no further action is taken.
+type configuringPhaseReconciler struct {
+	datastore  Datastore
+	kubeclient Kubeclient
+}
+
+func (r *configuringPhaseReconciler) Reconcile(ctx context.Context, in *v1alpha1.OperatorSource) (out *v1alpha1.OperatorSource, status *NextPhase, err error) {
+	logger := getLoggerWithOperatorSourceTypeFields(in)
+
+	out = in
+
+	cscName := getCatalogSourceConfigName(in.Name)
+	into := newCatalogSourceConfigTypeWithMetadata(in.Namespace, cscName)
+
+	err = r.kubeclient.Get(into)
+
+	if err == nil {
+		logger.Infof("No action taken, CatalogSourceConfig [name=%s] already exists", cscName)
+		status = getNexPhase(v1alpha1.OperatorSourcePhaseSucceeded)
+		return
+	}
+
+	if !k8s_errors.IsNotFound(err) {
+		status = getNexPhaseWithMessage(v1alpha1.OperatorSourcePhaseFailed, err.Error())
+		return
+	}
+
+	manifests := r.datastore.GetPackageIDs()
+	o := newCatalogSourceConfigType(getCatalogSourceConfigName(out.Name), manifests, out)
```

Comment:
I would prefer a new object in pristine state to start with. `into` is not guaranteed to be in pristine form after being used by `sdk.Get`. Also, re-purposing a local variable is something we want to avoid.

Reply:
I agree in this context as I missed that it was being passed into out. But re-purposing a local variable is something we want to avoid needn't be a general rule of thumb given we don't want to create unnecessary variables.